### PR TITLE
feat: end-to-end supervised fine-tuning for text classification

### PR DESCRIPTION
**Pull request by gabrielahrlr**
# Description

Thank you for opening a Pull Request!
Before submitting your PR, there are a few things you can do to make sure it goes smoothly:

- [ X] Follow the [`CONTRIBUTING` Guide](https://github.com/GoogleCloudPlatform/generative-ai/blob/main/CONTRIBUTING.md).
- [ X] You are listed as the author in your notebook or README file.
  - [ X] Your account is listed in [`CODEOWNERS`](https://github.com/GoogleCloudPlatform/generative-ai/blob/main/.github/CODEOWNERS) for the file(s).
- [ X] Make your Pull Request title in the <https://www.conventionalcommits.org/> specification.
- [ X] Ensure the tests and linter pass (Run `nox -s format` from the repository root to format).
- [ X] Appropriate docs were updated (if necessary)

Fixes #<issue_number_goes_here> 🦕

### CODE REVIEW

**Comment by holtskinner** [comment]
Is this PR supposed to be a replacement for #787 ?

**Comment by gabrielahrlr** [comment]
@holtskinner I went through all the errors from  lint code base and lint notebook errors and fixed them. The remaining ones make no sense to me. As per spelling errors, the errors are words likeICML (where dataset was published) last names from the authors of the dataset, automl and lopgrobs, which are not spelling errors.

**Comment by holtskinner** [comment]
I've added some of the terms to the spelling allowlist. You can see the remaining errors in the summary. It explains what the errors are and what they should be.

https://github.com/GoogleCloudPlatform/generative-ai/actions/runs/10254677291?pr=919

**Comment by gabrielahrlr** [comment]
Thanks Holt! All Spelling errors are now fixed. The Lint Notebook and Lint Code errors remain, as I have made different changes but nothing fix them.

File: gemini/tuning/gemini_supervised_fine-tuning_text_classification/gemini_supervised_finetuning_text_classification.ipynb [hunk 1]
```diff
@@ -0,0 +1,2724 @@
...
+        "In this step, we will load the raw data and create training, validation and test sets. Later these datasets will be used to perform  different types of adaptations to Gemini models for the task under consideration.\n"
+      ]
+    },
+    {
+      "cell_type": "markdown",
+      "metadata": {
+        "id": "2f475c9eceb9"
+      },
+      "source": [
+        "Load the dataset from HuggingFace"
```
Comment (holtskinner) [inline gemini/tuning/gemini_supervised_fine-tuning_text_classification/gemini_supervised_finetuning_text_classification.ipynb:553]:
```suggestion
        "Load the dataset from Hugging Face"
```

File: gemini/tuning/gemini_supervised_fine-tuning_text_classification/gemini_supervised_finetuning_text_classification.ipynb [hunk 2]
```diff
@@ -0,0 +1,2724 @@
...
+        "    if postprocessing:\n",
+        "        df[predictions_column] = df[predictions_column].apply(\n",
+        "            lambda x: predictions_postprocessing(x)\n",
+        "        )\n",
+        "\n",
+        "    metrics_report = classification_report(\n",
+        "        df[target_column], df[predictions_column], output_dict=True\n",
+        "    )\n",
+        "    overall_macro_f1_score = f1_score(\n",
+        "        df[target_column], df[predictions_column], average=\"macro\"\n",
```
Comment (holtskinner) [inline gemini/tuning/gemini_supervised_fine-tuning_text_classification/gemini_supervised_finetuning_text_classification.ipynb:785]: These methods can be simplified, Example:

```py
def predictions_postprocessing(text: str) -> str:
    """
    Cleans the predicted class label string.

    Args:
        text (str): The predicted class label string.

    Returns:
        str: The cleaned class label string.
    """
    return text.strip().lower()

def evaluate_predictions(
    df: pd.DataFrame,
    target_column: str = "label_text",
    predictions_column: str = "predicted_labels",
    postprocessing: bool = True,
) -> Dict[str, float]:
    """ 
    Batch evaluation of predictions, returns a dictionary with the metric.
    
    Args:
       - df (pandas.DataFrame):  a pandas dataframe with two mandatory columns, a target column with
       the actual true values, and a predictions column with the predicted values.
       - target_column (str): column name with the actual ground truth values
       - predictions_column (str): column name with the model predictions
       - postprocessing (bool): whether to apply postprocessing to predictions.
    
    Returns:
        Dict[str, float]: Dictionary of evaluation metrics.
    """
    if postprocessing:
        df[predictions_column] = df[predictions_column].apply(predictions_postprocessing)

    y_true = df[target_column]
    y_pred = df[predictions_column]
    
    metrics_report = classification_report(y_true, y_pred, output_dict=True)
    overall_macro_f1_score = f1_score(y_true, y_pred, average="macro")
    overall_micro_f1_score = f1_score(y_true, y_pred, average="micro")
    weighted_precision = precision_score(y_true, y_pred, average="weighted")
    weighted_recall = recall_score(y_true, y_pred, average="weighted")

    metrics = {
        "accuracy": metrics_report["accuracy"],
        "weighted precision": weighted_precision,
        "weighted recall": weighted_recall,
        "macro f1": overall_macro_f1_score,
        "micro f1": overall_micro_f1_score,
    }

    categories = ["business", "sport", "politics", "tech", "entertainment"]
    for category in categories:
        if category in metrics_report:
            metrics[f"{category}_f1_score"] = metrics_report[category]["f1-score"]

    return metrics
```

File: gemini/tuning/gemini_supervised_fine-tuning_text_classification/gemini_supervised_finetuning_text_classification.ipynb [hunk 3]
```diff
@@ -0,0 +1,2724 @@
...
+    },
+    {
+      "cell_type": "code",
+      "execution_count": null,
+      "metadata": {
+        "id": "OLitVLOFqPQe"
+      },
+      "outputs": [],
+      "source": [
+        "generation_config = {\"max_output_tokens\": 10, \"temperature\": 0}\n",
```
Comment (holtskinner) [inline gemini/tuning/gemini_supervised_fine-tuning_text_classification/gemini_supervised_finetuning_text_classification.ipynb:1023]: Please use the `GenerationConfig` class. I'd also recommend just re-using the same `generation_config` and `safety_settings` throughout the tests if possible.

**Review by holtskinner — APPROVED** [review]
Just a few more suggestions for improvements

**Comment by holtskinner** [comment]
I would also recommend putting the methods in the python file in the notebook itself so the users won't need to download another file when using colab.

**Comment by gabrielahrlr** [comment]
@holtskinner  fixed all errors and refactored as per suggestions. Erwin has recommended to have the 
"migration" functions separately for portability, but overall I agree that having inside the same colab is easier, so just put them back in the notebook as well and deleted the separate file.

Many thanks for all the great feedback. LMK if sth else needs to be changed to finally merge it into main :)

**Pull request merged**